### ISSUE TITLE
json_reader, json_parser, csv_parser, csv_reader use std::allocator only

### ISSUE THREAD

**Issue by xezon** [danielaparker/jsoncons#134]
This is no bug. Rather a feature request. Considering the json values respect custom allocators, it is odd that supporting classes don't use custom allocators too.

**Comment by danielaparker** [comment]
Agreed :-) The reason is that you were the first to ask. I'll look into it.

Daniel

**Comment by risksnap** [comment]
These classes (and a few more) now have an Allocator template parameter that defaults to std::allocator<CharT>, to allocate temporary memory storage used in the production of a json value. Only stateless allocators are currently supported for these classes, support for stateful allocators will be added later. 

Example:

#include <boost/pool/pool_alloc.hpp>
#include <jsoncons/json.hpp>

```c++
int main()
{
    jsoncons::json_decoder<json,boost::pool_allocator<char>> decoder;
    std::string s("[1,2,3,4,5,6]");
    std::istringstream is(s);

    basic_json_reader<char,boost::pool_allocator<char>> reader(is,decoder); 
    reader.read();

    json j = decoder.get_result();
}
```

**Comment by xezon** [comment]
```c++
template <class CharT,class Allocator=std::allocator<CharT>>
class basic_csv_parameters
{
    typedef std::basic_string<CharT,std::char_traits<CharT>,Allocator> string_type;

    std::vector<string_type,Allocator> column_names_;
    std::vector<detail::csv_type_info,Allocator> column_types_;
    std::vector<string_type,Allocator> column_defaults_;
}
```

This ain't right and won't compile in MSVC c++17

An allocator cannot have a mismatching type.

**Comment by danielaparker** [comment]
You're right, for the standard library containers, e.g. vector<T,Allocator>, it's undefined behavior if Allocator::value_type is not the same as T. I knew that, but forgot :-(.  

All the implementations that I've tested with use `rebind_alloc` to rebind the Allocator template parameter to the container value type, and don't give a diagnostic, so I missed that when I checked in this code.    

I've changed the code to rebind the Allocator template parameter to the value_type for the standard library containers wherever they appear in the code.
